src/crypto.rs

```diff
@@ -0,0 +1,217 @@
+// Copyright 2018 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under the MIT license <LICENSE-MIT
+// http://opensource.org/licenses/MIT> or the Modified BSD license <LICENSE-BSD
+// https://opensource.org/licenses/BSD-3-Clause>, at your option. This file may not be copied,
+// modified, or distributed except according to those terms. Please review the Licences for the
+// specific language governing permissions and limitations relating to use of the SAFE Network
+// Software.
+
+use maidsafe_utilities::serialisation::{deserialise, serialise};
+use safe_crypto::{PublicEncryptKey, SecretEncryptKey, SharedSecretKey};
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+
+/// Simplifies encryption by holding the necessary context - encryption keys.
+/// Allows "null" encryption where data is only serialized. See: null object pattern.
+#[derive(Clone, Debug)]
+pub enum EncryptContext {
+    /// No encryption.
+    Null,
+    /// Encryption + authentication
+    Authenticated { shared_key: SharedSecretKey },
+    /// No message authentication. Only encrypt operation is allowed.
+    AnonymousEncrypt {
+        /// Their public key.
+        their_pk: PublicEncryptKey,
+    },
+}
+
+impl Default for EncryptContext {
+    /// Default is "null" encryption.
+    fn default() -> Self {
+        Self::null()
+    }
+}
+
+impl EncryptContext {
+    /// Contructs "null" encryption context which actually does no encryption.
+    /// In this case data is simply serialized but not encrypted.
+    pub fn null() -> Self {
+        EncryptContext::Null
+    }
+
+    /// Construct crypto context that encrypts and authenticate messages.
+    pub fn authenticated(shared_key: SharedSecretKey) -> Self {
+        EncryptContext::Authenticated { shared_key }
+    }
+
+    /// Constructs crypto context that is only meant for unauthenticated encryption.
+    pub fn anonymous_encrypt(their_pk: PublicEncryptKey) -> Self {
+        EncryptContext::AnonymousEncrypt { their_pk }
+    }
+
+    /// Serialize given structure and encrypt it.
+    pub fn encrypt<T: Serialize>(&self, msg: &T) -> ::Res<Vec<u8>> {
+        Ok(match *self {
+            EncryptContext::Null => serialise(msg)?,
```

Comment:
Do we need to waste resources serialising/deser if no crypto context exists ? What if we return data as is for those cases ?

Reply:
We'll we need to retain the same API and given any object return `Vec<u8>`